File: Libraries/ART/ARTGroup.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <React/ARTGroup.h>

@implementation ARTGroup

- (void)renderLayerTo:(CGContextRef)context
{
  if (!CGRectIsEmpty(self.clipping)) {
    CGContextClipToRect(context, self.clipping);
  }

  for (UIView *subview in self.subviews) {
    if ([subview respondsToSelector:@selector(renderTo:)]) {
      [(ARTNode *)subview renderTo:context];
    } else {
      // This is needed for legacy interop layer. Legacy interop layer
      // is superview of the view that it is bridging, that's why we need
      // to grab its first subview.
      [(ARTNode *)subview.subviews.firstObject renderTo:context];
    }
  }
}

@end

File: ReactCommon/fabric/components/art/state/ARTGroup.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <react/components/art/ARTGroup.h>
#include <react/components/art/ARTElement.h>
#include <react/components/art/conversions.h>

namespace facebook {
namespace react {

bool ARTGroup::operator==(const ARTElement &rhs) const {
  if (rhs.elementType != ARTElementType::Group) {
    return false;
  }
  auto group = (const ARTGroup &)(rhs);
  return std::tie(elementType, opacity, transform, clipping) ==
      std::tie(
             group.elementType,
             group.opacity,
             group.transform,
             group.clipping) &&
      elements == group.elements;
}

bool ARTGroup::operator!=(const ARTElement &rhs) const {
  return !(*this == rhs);
}

#ifdef ANDROID
folly::dynamic ARTGroup::getDynamic() const {
  return toDynamic(*this);
}
#endif

} // namespace react
} // namespace facebook
